File: .UNSAFE-ALL-BINDINGS-DELETED-FINAL-20260125/bindings/scala/src/main/scala/proven/SafeDateTime.scala
// SPDX-License-Identifier: PMPL-1.0

package proven

import scala.util.Try

/**
 * Parsed date components.
 */
case class Date(year: Int, month: Int, day: Int):
  require(month >= 1 && month <= 12, "Month must be 1-12")
  require(day >= 1 && day <= SafeDateTime.daysInMonth(year, month).getOrElse(31), "Invalid day for month")

  /**
   * Format as ISO 8601 date (YYYY-MM-DD).
   */
  def formatIso: String = f"$year%04d-$month%02d-$day%02d"

  /**
   * Check if this date is a weekend.
   */
  def isWeekend: Boolean =
    val dayOfWeek = SafeDateTime.dayOfWeek(year, month, day)
    dayOfWeek == 0 || dayOfWeek == 6

  /**
   * Add days to this date.
   */
  def plusDays(days: Int): Date =
    SafeDateTime.addDays(this, days)

  override def toString: String = formatIso

/**
 * Parsed time components.
 */
case class Time(hour: Int, minute: Int, second: Int, nanos: Int = 0):
  require(hour >= 0 && hour <= 23, "Hour must be 0-23")
  require(minute >= 0 && minute <= 59, "Minute must be 0-59")
  require(second >= 0 && second <= 59, "Second must be 0-59")
  require(nanos >= 0 && nanos < 1000000000, "Nanos must be 0-999999999")

  /**
   * Format as ISO 8601 time (HH:MM:SS).
   */
  def formatIso: String =
    if nanos == 0 then f"$hour%02d:$minute%02d:$second%02d"
    else f"$hour%02d:$minute%02d:$second%02d.${nanos}%09d".stripSuffix("0" * (9 - nanos.toString.length))

  /**
   * Convert to total seconds since midnight.
   */
  def toSeconds: Int = hour * 3600 + minute * 60 + second

  override def toString: String = formatIso

/**
 * Combined date and time.
 */
case class DateTime(date: Date, time: Time):
  /**
   * Format as ISO 8601 datetime.
   */
  def formatIso: String = s"${date.formatIso}T${time.formatIso}"

  override def toString: String = formatIso

/**
 * Safe datetime parsing and validation.
 */
object SafeDateTime:

  /**
   * Check if a year is a leap year.
   */
  def isLeapYear(year: Int): Boolean =
    (year % 4 == 0 && year % 100 != 0) || (year % 400 == 0)

  /**
   * Get the number of days in a month.
   */
  def daysInMonth(year: Int, month: Int): Option[Int] = month match
    case 1 | 3 | 5 | 7 | 8 | 10 | 12 => Some(31)
    case 4 | 6 | 9 | 11 => Some(30)
    case 2 => Some(if isLeapYear(year) then 29 else 28)
    case _ => None

  /**
   * Validate and create a date.
   */
  def validateDate(year: Int, month: Int, day: Int): Either[String, Date] =
    if month < 1 || month > 12 then
      Left(s"Invalid month: $month")
    else
      daysInMonth(year, month) match
        case Some(maxDay) if day >= 1 && day <= maxDay =>
          Right(Date(year, month, day))
        case Some(maxDay) =>
          Left(s"Invalid day $day for month $month (max: $maxDay)")
        case None =>
          Left(s"Invalid month: $month")

  /**
   * Validate and create a time.
   */
  def validateTime(hour: Int, minute: Int, second: Int): Either[String, Time] =
    if hour < 0 || hour > 23 then Left(s"Invalid hour: $hour")
    else if minute < 0 || minute > 59 then Left(s"Invalid minute: $minute")
    else if second < 0 || second > 59 then Left(s"Invalid second: $second")
    else Right(Time(hour, minute, second))

  /**
   * Parse ISO 8601 date (YYYY-MM-DD).
   */
  def parseIsoDate(s: String): Either[String, Date] =
    val parts = s.trim.split("-")
    if parts.length != 3 then
      Left("Invalid date format, expected YYYY-MM-DD")
    else
      for
        year <- Try(parts(0).toInt).toEither.left.map(_ => "Invalid year")
        month <- Try(parts(1).toInt).toEither.left.map(_ => "Invalid month")
        day <- Try(parts(2).toInt).toEither.left.map(_ => "Invalid day")
        date <- validateDate(year, month, day)
      yield date

  /**
   * Parse ISO 8601 time (HH:MM:SS or HH:MM:SS.nnn).
   */
  def parseIsoTime(s: String): Either[String, Time] =
    val (timePart, nanos) = s.trim.indexOf('.') match
      case -1 => (s.trim, 0)
      case i =>
        val nanosStr = s.substring(i + 1).padTo(9, '0').take(9)
        (s.substring(0, i), Try(nanosStr.toInt).getOrElse(0))

    val parts = timePart.split(":")
    if parts.length < 2 || parts.length > 3 then
      Left("Invalid time format, expected HH:MM or HH:MM:SS")
    else
      for
        hour <- Try(parts(0).toInt).toEither.left.map(_ => "Invalid hour")
        minute <- Try(parts(1).toInt).toEither.left.map(_ => "Invalid minute")
        second <- if parts.length > 2 then
          Try(parts(2).toInt).toEither.left.map(_ => "Invalid second")
        else Right(0)
        time <- validateTime(hour, minute, second).map(_.copy(nanos = nanos))
      yield time

  /**
   * Parse ISO 8601 datetime (YYYY-MM-DDTHH:MM:SS).
   */
  def parseIsoDateTime(s: String): Either[String, DateTime] =
    val parts = s.trim.split("[T ]")
    if parts.length != 2 then
      Left("Invalid datetime format, expected YYYY-MM-DDTHH:MM:SS")
    else
      for
        date <- parseIsoDate(parts(0))
        time <- parseIsoTime(parts(1).takeWhile(c => c != 'Z' && c != '+' && c != '-'))
      yield DateTime(date, time)

  /**
   * Get day of week (0 = Sunday, 6 = Saturday) using Zeller's congruence.
   */
  def dayOfWeek(year: Int, month: Int, day: Int): Int =
    val m = if month < 3 then month + 12 else month
    val y = if month < 3 then year - 1 else year
    val k = y % 100
    val j = y / 100
    val h = (day + (13 * (m + 1)) / 5 + k + k / 4 + j / 4 - 2 * j) % 7
    ((h + 6) % 7) // Convert to Sunday = 0

  /**
   * Add days to a date.
   */
  def addDays(date: Date, days: Int): Date =
    var year = date.year
    var month = date.month
    var day = date.day + days

    while day > daysInMonth(year, month).getOrElse(28) do
      day -= daysInMonth(year, month).getOrElse(28)
      month += 1
      if month > 12 then
        month = 1
        year += 1

    while day < 1 do
      month -= 1
      if month < 1 then
        month = 12
        year -= 1
      day += daysInMonth(year, month).getOrElse(28)

    Date(year, month, day)

  /**
   * Calculate days between two dates.
   */
  def daysBetween(d1: Date, d2: Date): Int =
    toJulianDay(d2) - toJulianDay(d1)

  /**
   * Convert to Julian day number for calculations.
   */
  private def toJulianDay(date: Date): Int =
    val a = (14 - date.month) / 12
    val y = date.year + 4800 - a
    val m = date.month + 12 * a - 3
    date.day + (153 * m + 2) / 5 + 365 * y + y / 4 - y / 100 + y / 400 - 32045

  /**
   * Check if a date is valid.
   */
  def isValidDate(year: Int, month: Int, day: Int): Boolean =
    validateDate(year, month, day).isRight

  /**
   * Check if a time is valid.
   */
  def isValidTime(hour: Int, minute: Int, second: Int): Boolean =
    validateTime(hour, minute, second).isRight

  /**
   * Get the current date (system timezone).
   */
  def today: Date =
    val now = java.time.LocalDate.now()
    Date(now.getYear, now.getMonthValue, now.getDayOfMonth)

  /**
   * Get the current time (system timezone).
   */
  def now: Time =
    val t = java.time.LocalTime.now()
    Time(t.getHour, t.getMinute, t.getSecond, t.getNano)

File: .UNSAFE-ALL-BINDINGS-DELETED-FINAL-20260125/bindings/php/src/Proven/SafeDateTime.php
<?php
// SPDX-License-Identifier: PMPL-1.0
// SPDX-FileCopyrightText: 2025 Hyperpolymath

declare(strict_types=1);

namespace Proven;

/**
 * Parsed date components.
 */
readonly class Date
{
    public function __construct(
        /** Year (e.g., 2025) */
        public int $year,
        /** Month (1-12) */
        public int $month,
        /** Day (1-31) */
        public int $day,
    ) {}

    /**
     * Format as ISO 8601 date string.
     */
    public function format(): string
    {
        return sprintf('%04d-%02d-%02d', $this->year, $this->month, $this->day);
    }

    /**
     * Get the day of week (1=Monday, 7=Sunday).
     */
    public function dayOfWeek(): int
    {
        $date = \DateTimeImmutable::createFromFormat('Y-m-d', $this->format());
        if ($date === false) {
            return 1;
        }
        return (int)$date->format('N');
    }

    /**
     * Get the day of year (1-366).
     */
    public function dayOfYear(): int
    {
        $date = \DateTimeImmutable::createFromFormat('Y-m-d', $this->format());
        if ($date === false) {
            return 1;
        }
        return (int)$date->format('z') + 1;
    }

    /**
     * Check if this date is in a leap year.
     */
    public function isLeapYear(): bool
    {
        return SafeDateTime::isLeapYear($this->year);
    }

    public function __toString(): string
    {
        return $this->format();
    }
}

/**
 * Parsed time components.
 */
readonly class Time
{
    public function __construct(
        /** Hour (0-23) */
        public int $hour,
        /** Minute (0-59) */
        public int $minute,
        /** Second (0-59) */
        public int $second,
        /** Nanoseconds (0-999999999) */
        public int $nanos = 0,
    ) {}

    /**
     * Format as ISO 8601 time string.
     */
    public function format(bool $includeNanos = false): string
    {
        $base = sprintf('%02d:%02d:%02d', $this->hour, $this->minute, $this->second);
        if ($includeNanos && $this->nanos > 0) {
            $base .= '.' . sprintf('%09d', $this->nanos);
        }
        return $base;
    }

    /**
     * Get total seconds since midnight.
     */
    public function toSeconds(): int
    {
        return $this->hour * 3600 + $this->minute * 60 + $this->second;
    }

    public function __toString(): string
    {
        return $this->format();
    }
}

/**
 * Combined date and time.
 */
readonly class DateTime
{
    public function __construct(
        public Date $date,
        public Time $time,
        /** Timezone offset in minutes (e.g., -300 for UTC-5) */
        public ?int $tzOffset = null,
    ) {}

    /**
     * Format as ISO 8601 datetime string.
     */
    public function format(): string
    {
        $result = $this->date->format() . 'T' . $this->time->format();

        if ($this->tzOffset !== null) {
            if ($this->tzOffset === 0) {
                $result .= 'Z';
            } else {
                $sign = $this->tzOffset >= 0 ? '+' : '-';
                $absOffset = abs($this->tzOffset);
                $hours = intdiv($absOffset, 60);
                $minutes = $absOffset % 60;
                $result .= sprintf('%s%02d:%02d', $sign, $hours, $minutes);
            }
        }

        return $result;
    }

    /**
     * Convert to Unix timestamp.
     */
    public function toTimestamp(): int
    {
        $dt = new \DateTimeImmutable($this->format());
        return $dt->getTimestamp();
    }

    public function __toString(): string
    {
        return $this->format();
    }
}

/**
 * Safe datetime parsing and validation.
 *
 * Provides ISO 8601 compliant datetime parsing with proper validation.
 */
class SafeDateTime
{
    /**
     * Check if a year is a leap year.
     */
    public static function isLeapYear(int $year): bool
    {
        return ($year % 4 === 0 && $year % 100 !== 0) || ($year % 400 === 0);
    }

    /**
     * Get the number of days in a month.
     *
     * @return Option<int>
     */
    public static function daysInMonth(int $year, int $month): Option
    {
        return match ($month) {
            1, 3, 5, 7, 8, 10, 12 => Option::some(31),
            4, 6, 9, 11 => Option::some(30),
            2 => Option::some(self::isLeapYear($year) ? 29 : 28),
            default => Option::none(),
        };
    }

    /**
     * Validate and create a date.
     *
     * @return Result<Date>
     */
    public static function validateDate(int $year, int $month, int $day): Result
    {
        if ($month < 1 || $month > 12) {
            return Result::err(ProvenError::validationError("Invalid month: $month"));
        }

        $maxDay = self::daysInMonth($year, $month)->unwrapOr(0);
        if ($day < 1 || $day > $maxDay) {
            return Result::err(ProvenError::validationError("Invalid day $day for month $month"));
        }

        return Result::ok(new Date($year, $month, $day));
    }

    /**
     * Validate and create a time.
     *
     * @return Result<Time>
     */
    public static function validateTime(int $hour, int $minute, int $second, int $nanos = 0): Result
    {
        if ($hour < 0 || $hour > 23) {
            return Result::err(ProvenError::validationError("Invalid hour: $hour"));
        }
        if ($minute < 0 || $minute > 59) {
            return Result::err(ProvenError::validationError("Invalid minute: $minute"));
        }
        if ($second < 0 || $second > 59) {
            return Result::err(ProvenError::validationError("Invalid second: $second"));
        }
        if ($nanos < 0 || $nanos > 999999999) {
            return Result::err(ProvenError::validationError("Invalid nanoseconds: $nanos"));
        }

        return Result::ok(new Time($hour, $minute, $second, $nanos));
    }

    /**
     * Parse an ISO 8601 date (YYYY-MM-DD).
     *
     * @return Result<Date>
     */
    public static function parseIsoDate(string $s): Result
    {
        $s = trim($s);
        $parts = explode('-', $s);

        if (count($parts) !== 3) {
            return Result::err(ProvenError::parseError('Invalid date format, expected YYYY-MM-DD'));
        }

        $year = filter_var($parts[0], FILTER_VALIDATE_INT);
        $month = filter_var($parts[1], FILTER_VALIDATE_INT);
        $day = filter_var($parts[2], FILTER_VALIDATE_INT);

        if ($year === false || $month === false || $day === false) {
            return Result::err(ProvenError::parseError('Invalid date components'));
        }

        return self::validateDate($year, $month, $day);
    }

    /**
     * Parse an ISO 8601 time (HH:MM:SS or HH:MM:SS.nnn).
     *
     * @return Result<Time>
     */
    public static function parseIsoTime(string $s): Result
    {
        $s = trim($s);

        // Handle fractional seconds
        $nanos = 0;
        $dotPos = strpos($s, '.');
        if ($dotPos !== false) {
            $fracPart = substr($s, $dotPos + 1);
            $s = substr($s, 0, $dotPos);
            $fracPart = str_pad(substr($fracPart, 0, 9), 9, '0');
            $nanos = (int)$fracPart;
        }

        $parts = explode(':', $s);
        if (count($parts) < 2 || count($parts) > 3) {
            return Result::err(ProvenError::parseError('Invalid time format, expected HH:MM or HH:MM:SS'));
        }

        $hour = filter_var($parts[0], FILTER_VALIDATE_INT);
        $minute = filter_var($parts[1], FILTER_VALIDATE_INT);
        $second = count($parts) > 2 ? filter_var($parts[2], FILTER_VALIDATE_INT) : 0;

        if ($hour === false || $minute === false || $second === false) {
            return Result::err(ProvenError::parseError('Invalid time components'));
        }

        return self::validateTime($hour, $minute, $second, $nanos);
    }

    /**
     * Parse an ISO 8601 datetime (YYYY-MM-DDTHH:MM:SS[.nnn][Z|+HH:MM]).
     *
     * @return Result<DateTime>
     */
    public static function parseIsoDateTime(string $s): Result
    {
        $s = trim($s);

        // Handle timezone
        $tzOffset = null;
        if (str_ends_with($s, 'Z')) {
            $tzOffset = 0;
            $s = substr($s, 0, -1);
        } elseif (preg_match('/([+-])(\d{2}):?(\d{2})$/', $s, $matches)) {
            $sign = $matches[1] === '+' ? 1 : -1;
            $hours = (int)$matches[2];
            $minutes = (int)$matches[3];
            $tzOffset = $sign * ($hours * 60 + $minutes);
            $s = substr($s, 0, -strlen($matches[0]));
        }

        // Split date and time
        $tPos = stripos($s, 'T');
        if ($tPos === false) {
            $tPos = strpos($s, ' ');
        }

        if ($tPos === false) {
            return Result::err(ProvenError::parseError('Invalid datetime format, expected YYYY-MM-DDTHH:MM:SS'));
        }

        $datePart = substr($s, 0, $tPos);
        $timePart = substr($s, $tPos + 1);

        $dateResult = self::parseIsoDate($datePart);
        if ($dateResult->isErr()) {
            return $dateResult;
        }

        $timeResult = self::parseIsoTime($timePart);
        if ($timeResult->isErr()) {
            return $timeResult;
        }

        return Result::ok(new DateTime(
            $dateResult->unwrap(),
            $timeResult->unwrap(),
            $tzOffset,
        ));
    }

    /**
     * Get the current date.
     */
    public static function today(): Date
    {
        $now = new \DateTimeImmutable();
        return new Date(
            (int)$now->format('Y'),
            (int)$now->format('n'),
            (int)$now->format('j'),
        );
    }

    /**
     * Get the current time.
     */
    public static function now(): Time
    {
        $now = new \DateTimeImmutable();
        return new Time(
            (int)$now->format('G'),
            (int)$now->format('i'),
            (int)$now->format('s'),
            (int)$now->format('u') * 1000,
        );
    }

    /**
     * Get the current datetime.
     */
    public static function nowDateTime(): DateTime
    {
        $now = new \DateTimeImmutable();
        return new DateTime(
            self::today(),
            self::now(),
            (int)($now->getOffset() / 60),
        );
    }

    /**
     * Create a DateTime from a Unix timestamp.
     */
    public static function fromTimestamp(int $timestamp, ?int $tzOffset = null): DateTime
    {
        $dt = (new \DateTimeImmutable())->setTimestamp($timestamp);

        return new DateTime(
            new Date(
                (int)$dt->format('Y'),
                (int)$dt->format('n'),
                (int)$dt->format('j'),
            ),
            new Time(
                (int)$dt->format('G'),
                (int)$dt->format('i'),
                (int)$dt->format('s'),
            ),
            $tzOffset,
        );
    }

    /**
     * Add days to a date.
     */
    public static function addDays(Date $date, int $days): Date
    {
        $dt = \DateTimeImmutable::createFromFormat('Y-m-d', $date->format());
        if ($dt === false) {
            return $date;
        }
        $dt = $dt->modify("$days days");
        return new Date(
            (int)$dt->format('Y'),
            (int)$dt->format('n'),
            (int)$dt->format('j'),
        );
    }

    /**
     * Calculate the difference in days between two dates.
     */
    public static function diffDays(Date $from, Date $to): int
    {
        $dtFrom = \DateTimeImmutable::createFromFormat('Y-m-d', $from->format());
        $dtTo = \DateTimeImmutable::createFromFormat('Y-m-d', $to->format());

        if ($dtFrom === false || $dtTo === false) {
            return 0;
        }

        $diff = $dtTo->diff($dtFrom);
        return $diff->invert ? $diff->days : -$diff->days;
    }
}
